feat: Bootstrap 5

jsit opened the pull request:
Fully remove Bootstrap 4 dependency in favor of Bootstrap 5

I should note that this changes the main green color in the Litely theme; Bootstrap 5 automatically generates foreground colors for given background colors, including with a set minimum contrast level.

I set the minimum contrast level to 3 (which passes [WCAG AA](https://webaim.org/resources/contrastchecker/) for "large" text) -- the Bootstrap 5 default is a contrast level of 4.5, which would require a much darker green to handle. (The current white-on-green is only 2.23.)

Should probably offer a high-contrast theme that gets automatically selected for the `prefers-contrast: more` media query.

cc @fheft

ayan4m1 commented:
LGTM, thanks for this! 🚀

fheft commented:
Couldn't try it out yet, but code lgtm. Thanks for doing this! :rocket:

alectrocute commented:
I spent some time testing out this branch. Great work as always @jsit, just a few things that may or may not be related to these changes, but nonetheless, would be nice to have addressed –

- Error toasters have no background:

<img width="440" alt="Screenshot 2023-06-19 at 8 13 28 AM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/fc2fed29-5c71-428e-a40b-dc249ba97b3c">

- 'Delete Account' button has no background on darkly red:

<img width="1385" alt="Screenshot 2023-06-19 at 8 14 33 AM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/4e56da7a-e3e5-4017-a1b5-7d2057289e09">

- I wish we could fit all these on the same row and avoid wrap:

<img width="598" alt="Screenshot 2023-06-19 at 8 15 22 AM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/127258d2-5d19-442e-9301-9938e3dbba26">

- Native radio buttons on these shouldn't be visible:

<img width="1181" alt="Screenshot 2023-06-19 at 8 16 13 AM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/e3927dfa-38f4-47fe-af5f-9f204f4ef067">

- Same dealio here:

<img width="647" alt="Screenshot 2023-06-19 at 8 16 34 AM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/d0a8cce5-cd03-41cb-96b2-49e4b0be250d">

- Where thou `mr-2`:

<img width="1200" alt="Screenshot 2023-06-19 at 8 16 51 AM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/4912cda7-db3f-4fcf-bf8a-ac3523f4c01d">

- We can probably fit this on the same row:

<img width="1171" alt="Screenshot 2023-06-19 at 8 17 38 AM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/6a53b1b5-c534-412c-bf36-1bbb176fa78a">

alectrocute commented:
One more thing. When I run `yarn themes:build`, I get this error (and CSS) output:

<img width="1506" alt="Screenshot 2023-06-19 at 8 24 36 AM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/cb2c3f5d-5523-43b7-bf03-0e61f152a773">

Also, I'm wondering if we need to have `themes:build` ran upon `postinstall`? Are we committing the dist outputs instead?

jsit commented:
@alectrocute Thank you for the thorough review! Those issues should be addressed and I'll try to get to them this afternoon.

jsit commented:
@alectrocute All these issues should now be fixed.

fheft commented:
@jsit alectrocute and you did a great job, I haven't found any issues :slightly_smiling_face:

alectrocute commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -144,10 +144,10 @@ export class MarkdownTextArea extends Component<
           }
         />
-        <div className="form-group row">
+        <div className="input-group mb-3 row">
```
`input-group` has some side effects.

With `input-group`:

<img width="1151" alt="Screenshot 2023-06-19 at 5 50 24 PM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/e672d7b3-5829-4f7b-b5a4-1582c9155621">

Without:

<img width="1141" alt="Screenshot 2023-06-19 at 5 50 15 PM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/5c7d97d8-1c52-42d0-89a6-d96fe5faea0e">

Note how this `MarkdownTextarea` doesn't scale to the full width of the column.

alectrocute replied:
I also seriously wish we could get all those buttons on one line. Maybe we could put the help icon next to the label instead? 🤔

alectrocute replied:
If we could fit this all on one row, it would be _ perfection _.

jsit replied:
Ah, I think this is because I have some instances of `input-group` and `row` classes on the same element; it seems not to like that.

jsit replied:
> If we could fit all these icons all on one row, and have the textarea properly fit the space, it would be _ perfection 

Also remember, even if it fit, it would barely fit, so for anybody who increased their browser zoom or were on mobile, it would still wrap. It's a tight fit. I think it wraps kind of nicely, it doesn't bother me too too much.

alectrocute commented:
One more thing – Noticed some missing right margin on various items on `Modlog`:

<img width="1160" alt="Screenshot 2023-06-19 at 5 55 17 PM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/c23cee2e-bf64-40c2-ba34-e6f047a6ce66">

Anyway, pretty absurd how much better everything is looking/feeling compared to 0.17. Great work. Keep going!

alectrocute commented:
Tabs have a little bit of bottom border radius wonkiness:

<img width="587" alt="Screenshot 2023-06-19 at 5 58 12 PM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/fc94a04d-5e71-4b4b-927f-f3f229234642">

alectrocute commented:
Did this table background color change? It would be nice to have the background alternate light-dark, or have no background on the table rows at all (eg. how it's done in 0.17).

<img width="2056" alt="Screenshot 2023-06-19 at 6 00 22 PM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/3c5a487a-0110-44ab-90fc-55608a4c631d">

alectrocute commented:
At some point, our search form became a lot more squished in the corner. Would be nice to use all available space.

<img width="2056" alt="Screenshot 2023-06-19 at 6 03 14 PM" src="https://github.com/LemmyNet/lemmy-ui/assets/35377827/ee3b5a13-f81a-4a2d-b921-b029d3d95353">

alectrocute commented:
Low priority, but the tagline items in the admin panel have a strange hover background:

https://github.com/LemmyNet/lemmy-ui/assets/35377827/e02a1878-eecd-4de2-966b-c1054d13e43b

jsit commented:
> Tabs have a little bit of bottom border radius wonkiness:

@alectrocute These already look this way, and are fixed in #1382.

Thank you for all the feedback! Even in `main` there are some weird visual quirks, so it would be helpful if only regressions were pointed out here.

alectrocute commented:
Please ignore any feedback that isn't related to/shouldn't be addressed in this PR. Thanks @jsit, good work as always!

dessalines commented:
Thanks a ton for doing this! This should have some priority as the merge conflicts will start to pile up if we do too much work in other areas. 

@SleeplessOne1917 Do you think this should come before or after `0.18.0`?

SleeplessOne1917 commented:
@dessalines I think before 0.18 would be better.

jsit commented:
@dessalines @SleeplessOne1917 I'll work on getting these final issues sorted out. Thanks!

jsit commented:
@alectrocute 

> Low priority, but the tagline items in the admin panel have a strange hover background:

`main` does this too, not a regression.

> At some point, our search form became a lot more squished in the corner. Would be nice to use all available space.

Fixed.

> Tabs have a little bit of bottom border radius wonkiness:

This is in `main`; will be fixed in #1382.

> Did this table background color change?

Fixed.

> One more thing – Noticed some missing right margin on various items on Modlog:

Fixed.

> `input-group` has some side effects.

Fixed.

jsit commented:
> @SleeplessOne1917 Do you think this should come before or after `0.18.0`?

Personally I think it might be better to wait. This PR is pretty young for its size, 0.18 is already in RC status, and I’d hate to saddle any instances with any undetected regressions this might cause until the next release.

This isn’t solving much for the end user, just some dependency maintenance.

However you’re right about the merge conflicts potentially becoming a big problem, so it should probably be merged very shortly after 0.18 is cut at least.

But if it would be more work to wait, I understand that too.

dessalines commented:
I'll test this out locally rn.

Once we hit `0.18.0`, we should be able to do much faster releases. These bigger PRs always tend to be blockers, and then fixing the merge conflicts in them quickly becomes a pain, so I agree that it's good to get them in ASAP. 

It will probably be a bit on `0.18.0` anyway, as we're finding some back-end issues that need fixing. So I'm good merging this if most things test out okay.

dessalines reviewed (approved):
I tested it, and did a quick pass of some of the components, and everything looks good to me. 

Yall have been making my clunky UI into something so much better :) Thanks a ton.

I agree with sleepless and merge this as is, and any other regressions we find can be fixed in smaller more disectable PRs.
